### DISCUSSION TITLE
РК-2

### DISCUSSION THREAD

**Discussion by TonitaN** [TonitaN/TFL2021#11] [Announcements]
В каждом варианте будет описание трёх языков (словесное, свёрточное или в КЗ-форме с атрибутами). Как правило(!), один из этих языков - не КС, другой - КС, но не детерминированный, третий - DCFL. Будут и исключения, но только в сторону упрощения (т.е., например, два недетерминированных языка, но все КС; или два DCFL).

Подзадачи по каждому пункту:

1. Не КС-язык: обосновать это с помощью леммы о накачке (3 балла), привести дополнительное обоснование (использующее другой тип зависимости в накачке либо не использующее лемму о накачке) (2 балла); проанализировать, является ли язык DMFL (2 балла), либо проанализировать, является ли язык конъюнктивным (2 балла).
2. КС-язык, но не детерминированный: обосновать недетерминированность (3 балла), построить PDA или грамматику (2 балла), определить степень недетерминизма посредством исправляющих токенов (2 балла).
3. DCFL. Построить DPDA (2 балла), проверить на LL-свойства (2 балла), проверить на LR(0)-свойство (1 балл), проверить на VPL-свойство (2 балла). Тут кому-то может невероятно повезти и выпасть регулярный язык. В этом случае в проверке на LL-свойства потребуется привести LL(1)-грамматику, а проверка на VPL-свойства (тривиальная) заменяется на проверку на 1-однозначность.

Итог: за каждое задание можно получить максимум 7 баллов, всего максимум 21 балл. Однако, как легко заметить, в каждом задании есть особенно сложный пункт, который требует изучения вспомогательной теории. По основной теории расчёт на максимум 15 баллов.

**Comment by Ankalot** [comment]
Консультация сегодня будет в 18:00?

**Reply by TonitaN** [reply]
Всем удобно в 18-30?

**Reply by TonitaN** [reply]
Если не считать клоников, то 1 против 4 (4й голос "за" - я), поэтому сдвигаем на полчаса.

**Comment by TonitaN** [comment]
Для построения LR(0)-автоматов по вашим грамматикам берите Граммофон (или свои лабы, кто делает вариант 0).
Для тестирования LL(k)-свойства (если подозреваете, что язык LL(k) и построили грамматику, которая это должна обосновывать) - берите свои реализации first_k и follow_k и на их базе тестируйте. Есть проверяльщики LL(1)-свойства, но LL(2) и более - это уже ваша забота. Правда, если кто-то найдёт адекватный онлайн-инструмент, лучше напишите мне - я его добавлю в ссылки и протестирую. Вдруг на чём-то ломается - много что ломается в правильных руках, не только ваши лабы ^^

**Comment by boomb0om** [comment]
Добрый вечер!
Подскажите, а во сколько завтра будет консультация? И очно или заочно?

**Reply by TonitaN** [reply]
Заочно, ориентируемся на 16-30.

**Comment by TonitaN** [comment]
Есть две новости: хорошая и плохая. Плохая - похоже что сегодня с временем всё будет совсем плохо, и проводить консультацию будет 
 некогда (максимум полчаса - а это ни о чём). А хорошая - это вот:
https://github.com/TonitaN/FormalLanguageTheory/tree/main/2022/tasks
Обещанный разбор трёх пробных заданий здесь есть, поэтому я надеюсь, что ваши ожидания будут не слишком обмануты.

**Comment by TonitaN** [comment]
Кто ещё не видел, рекомендую на РК заглядывать сюда: 👍
https://github.com/theElusiveJoe/Obsidian-Vault

**Comment by bad-joker-bmstu** [comment]
Не уследил за коллегами, они ушли в архив и не вернулись ...
![Снимок экрана от 2022-12-09 23-14-43](https://user-images.githubusercontent.com/120220064/206791086-1c739d76-411b-4cf1-b847-4955796607a2.png)

![Снимок экрана от 2022-12-09 23-14-47](https://user-images.githubusercontent.com/120220064/206791039-0adee6d4-2a7b-42d0-880a-94dec6b684b3.png)

![Снимок экрана от 2022-12-09 23-14-53](https://user-images.githubusercontent.com/120220064/206791137-268c4934-c5c6-4e4b-b292-c9be747e8954.png)

![Снимок экрана от 2022-12-09 23-15-00](https://user-images.githubusercontent.com/120220064/206791184-a45aeccf-8d9c-4720-b02a-b5e4c04f7e28.png)
![Снимок экрана от 2022-12-09 23-15-08](https://user-images.githubusercontent.com/120220064/206791203-fdcdfd2d-38d2-4772-8a56-46e27c0ef31d.png)


![Снимок экрана от 2022-12-09 23-15-14](https://user-images.githubusercontent.com/120220064/206791222-6564af0b-0d29-4c9d-8a8d-2f737a57786a.png)
![Снимок экрана от 2022-12-09 23-15-30](https://user-images.githubusercontent.com/120220064/206791274-c62162f2-d5cf-405a-89d5-25d06e40d1f6.png)

![Снимок экрана от 2022-12-09 23-15-38](https://user-images.githubusercontent.com/120220064/206791328-cab6ee68-bed7-4223-8d7d-bb5ff4ea803c.png)

**Comment by TonitaN** [comment]
@mathhyyn, там действительно отрицательная накачка.

Возьмём слова `a a^(n+2) b a^(n) a`, `a a^(n+2) b a^n a b a^{2n+2}` и префикс `a a^(n+2) b a^n`. Если накачивать только его, то второе слово сразу же выйдет из языка при отрицательной накачке. Если же накачивать только суффикс `a^n` и ещё что-то за ним, то при положительной накачке выйдет из языка первое слово.
С первой частью слова с буквой `a` посередине я обошлась жестоко - минимизировала её до одной буквы `a`. Причём выбор более длинного слова на эту роль только ухудшил бы положение с `b` в суффиксах, т.к. оторвал бы часть слова от `u1`, но не от `u2`, а они и так почти равны по длине.

**Reply by mathhyyn** [reply]
самое грустное, что я точно такой же пример в черновике рассматривала(( но про отрицательную тогда не подумала
только там лучше `a a^(n+2) b a^n a b a^{2n+3}`

**Reply by TonitaN** [reply]
Ну если я его там найду, то балл будет, пусть и не полный. Засчитала же в своё время некую "жалкую регулярку, которая имеет хоть какое-то отношение к действительности")

**Reply by mathhyyn** [reply]
Черновик остался на руках, к сожалению

**Comment by TonitaN** [comment]
Народ, стартовое состояние в PDA надо как-то выделять: или номером, или стрелкой без начала. От автоматов-ловушек уже немного отвыкли, но теперь у всех в работах автоматы с непонятным стартовым )

**Comment by TonitaN** [comment]
Что-то сегодня очень грустное настроение у группы 51 :( У @Ankalot надо было всё-таки его работу отобрать - жалею, что не уследила, - всегда же лучше всех на парах задачи решал. Зато @xendalm отыгрался за прошлый раз.

**Comment by TonitaN** [comment]
По РК-2 результаты следующие:
- 1 место (13 баллов) @Lasadaf
- 2 место (12 баллов) @xendalm, @UsefulTornado, @KarumaRey
- 3 место (10 баллов) @p0rtale, @mathhyyn , @dak151449

Кого особенно жду на апелляцию (набрали по 5 баллов): @xtoter, @boomb0om , @StarikTenger, Полина Самохвалова (не помню аккаунт)
